src/actions/updateThoughts.ts

```diff
@@ -186,8 +188,34 @@ const updateThoughts = (
   const lexemeIndexOld = { ...state.thoughts.lexemeIndex }
   const lexemeIndexUpdatesOld = keyValueBy(lexemeIndexUpdates, key => ({ [key]: lexemeIndexOld[key] }))
 
-  // TODO: Can we use { overwritePending: !local } and get rid of the overwritePending option to updateThoughts? i.e. Are there any false positives when local is false?
-  const thoughtIndex = mergeUpdates(thoughtIndexOld, thoughtIndexUpdates, { overwritePending })
+  // Collect parentIds that may be affected (i.e., a child update could clear their pending flag)
+  const affectedParentIds = new Set<ThoughtId>()
+
+  const mergedUpdatesMap = keyValueBy(thoughtIndexUpdates, (id, thought) => {
+    if (thought) affectedParentIds.add(thought.parentId)
+    return { [id]: thought }
+  })
+
+  let thoughtIndex = mergeUpdates(thoughtIndexOld, mergedUpdatesMap, { overwritePending })
+
+  // Single sweep: clear pending from affected parents whose children are all loaded
+  if (!overwritePending) {
+    affectedParentIds.forEach(parentId => {
```

Comment:
We probably don't need `affectedParentIds` or `mergedUpdatesMap`. You can just loop through `thoughtIndexUpdates`. By splitting this simple loop + guard into multiple steps and variables, it is making the code appear more complicated than it actually is.

Reply:
Even better, could this logic be moved into `mergeUpdates`? It has access to `thoughtIndexOld`. That would be better encapsulated, if that's possible.